### PR TITLE
Limit VMResult cases when concrete

### PR DESCRIPTION
**Pull request by arcz**
## Description
This builds on top of https://github.com/ethereum/hevm/pull/427. Better types using `VMType` allowed to eliminate dead code and optimize concrete execution with specialized `branch`. The speedup is ~10% on ethereum-tests.

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

File: src/EVM.hs [hunk 1]
```diff
@@ -1251,32 +1251,11 @@ getCodeLocation vm = (vm.state.contract, vm.state.pc)
 query :: Query t s -> EVM t s ()
 query = assign #result . Just . HandleEffect . Query
 
-choose :: Choose t s -> EVM t s ()
+choose :: Choose s -> EVM Symbolic s ()
```
Comment (msooseth) [inline src/EVM.hs:1254]: Notes to self: so choose is only defined for Symbolic now.

File: src/EVM.hs [hunk 2]
```diff
@@ -2718,6 +2715,10 @@ instance VMOps Concrete where
 
   whenSymbolicElse _ a = a
 
+  partial _ = internalError "won't happen during concrete exec"
```
Comment (msooseth) [inline src/EVM.hs:2718]: Right, so `partial` cannot happen during concrete exec

File: src/EVM.hs [hunk 3]
```diff
@@ -2589,6 +2565,27 @@ instance VMOps Symbolic where
   toGas _ = ()
   whenSymbolicElse a _ = a
 
+  partial e = assign #result (Just (Unfinished e))
```
Comment (msooseth) [inline src/EVM.hs:2568]: `partial` now only needs to be defined for symbolic exec

File: src/EVM/Stepper.hs [hunk 4]
```diff
@@ -46,7 +45,7 @@ data Action t s a where
   Wait :: Query t s -> Action t s ()
 
   -- | Multiple things can happen
-  Ask :: Choose t s -> Action t s ()
+  Ask :: Choose s -> Action Symbolic s ()
```
Comment (msooseth) [inline src/EVM/Stepper.hs:48]: This is nice actually.

**Review by msooseth — APPROVED** [review]
Actually makes the code cleaner and easier to read, too. I like it a lot, thanks!

File: src/EVM.hs (hunk 3 above)
Comment (arcz) [reply src/EVM.hs:2568]: In the long run, `partial` should not be part of `VMOps` but it needs a slight refactor to other functions

**Pull request merged**